modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/service/ClientServiceInvokeRequest.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.ignite.internal.processors.platform.PlatformNativeException;
+import org.apache.ignite.internal.processors.platform.client.ClientConnectionContext;
+import org.apache.ignite.internal.processors.platform.client.ClientObjectResponse;
+import org.apache.ignite.internal.processors.platform.client.ClientRequest;
+import org.apache.ignite.internal.processors.platform.client.ClientResponse;
+import org.apache.ignite.internal.processors.platform.services.PlatformService;
+import org.apache.ignite.internal.processors.platform.services.PlatformServices;
+import org.apache.ignite.internal.processors.service.GridServiceProxy;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.services.Service;
+import org.apache.ignite.services.ServiceDescriptor;
+
+/**
+ * Request to invoke service method.
+ */
+public class ClientServiceInvokeRequest extends ClientRequest {
+    /** Flag keep binary mask. */
+    private static final byte FLAG_KEEP_BINARY_MASK = 0x01;
+
+    /** Methods cache. */
+    private static final Map<MethodDescriptor, Method> methodsCache = new ConcurrentHashMap<>();
+
+    /** Service name. */
+    private final String name;
+
+    /** Flags. */
+    private final byte flags;
+
+    /** Timeout. */
+    private final long timeout;
+
+    /** Nodes. */
+    private final Set<UUID> nodeIds;
+
+    /** Method name. */
+    private final String methodName;
+
+    /** Method parameter type IDs. */
+    private int[] paramTypeIds;
+
+    /** Service arguments. */
+    private final Object[] args;
+
+    /** Objects reader. */
+    private final BinaryRawReaderEx reader;
+
+    /**
+     * Constructor.
+     *
+     * @param reader Reader.
+     */
+    public ClientServiceInvokeRequest(BinaryReaderExImpl reader) {
+        super(reader);
+
+        name = reader.readString();
+
+        flags = reader.readByte();
+
+        timeout = reader.readLong();
+
+        int cnt = reader.readInt();
+
+        nodeIds = U.newHashSet(cnt);
+
+        for (int i = 0; i < cnt; i++)
+            nodeIds.add(reader.readUuid());
+
+        methodName = reader.readString();
+
+        paramTypeIds = reader.readIntArray();
+
+        int argCnt = reader.readInt();
+
+        args = new Object[argCnt];
+
+        // We can't deserialize some types (arrays of user defined types for example) from detached objects.
+        // On the other hand, deserialize should be done as part of process() call (not in constructor) for proper
+        // error handling.
+        // To overcome these issues we store binary reader reference, parse request in constructor (by reading detached
+        // objects), restore arguments starting position in input stream and deserialize arguments from input stream
+        // in process() method.
+        this.reader = reader;
```

Comment:
For example, if you marshal `Collections.singleton(new Person[] { person })` it will be unmarshalled as 'Collections.singleton(new Object[] { binaryObject })` by `readObjectDetached` and nothing we can do here to restore array parameter type.
It's not the only problem with detached objects. One more: you can't restore reference after `readObjectDetached`. For example, if you have `new Person[] { person, person }` (one object of "Person" class and two references to it), after unmarshalling and unwrapping binaries you will have two different objects of "Person" class.

Reply:
> `Collections.singleton`

This returns `Set`, which does not specific element type at runtime, so what's the problem?

> you can't restore reference after `readObjectDetached`

Well, this is a separate matter. For some reason we use `readObjectDetached` for two things at the same time: 
* To read as binary
* To read as detached (don't preserve references)

We can keep references if we want to.

Reply:
> so what's the problem?
It's not a problem to invoke the service, but can be a problem if some user-defined service checks element types. For example, a user can write something like `for (Object item : collection) if (item instance of Person[]) ...` and such a check will fail.
> We can keep references if we want to.
We can't. If we read two objects as detached there is no information inside these objects that it was the same object before marshaling.

Reply:
> If we read two objects as detached
Don't read as detached?

Reply:
To summarize - I don't strongly object to this. The code is a bit dirty and fragile, but the problem is localized within a single class. So I leave this up to you.

Reply:
> Don't read as detached?

We have two options: `readObject` and` readObjectDetached`. In case of `readObject` we will deserialize in the constructor, it's undesirable.